src/common/src/array/stream_chunk.rs

```diff
@@ -125,26 +126,20 @@ impl StreamChunk {
     }
 
     /// Build a `StreamChunk` from rows.
-    // TODO: introducing something like `StreamChunkBuilder` maybe better.
+    ///
+    /// Panics if the `rows` is empty.
+    ///
+    /// Should prefer using [`StreamChunkBuilder`] instead to avoid unnecessary
+    /// allocation of rows.
     pub fn from_rows(rows: &[(Op, impl Row)], data_types: &[DataType]) -> Self {
-        let mut array_builders = data_types
-            .iter()
-            .map(|data_type| data_type.create_array_builder(rows.len()))
-            .collect::<Vec<_>>();
-        let mut ops = vec![];
+        let mut builder = StreamChunkBuilder::new(rows.len() + 1, data_types.to_vec());
 
         for (op, row) in rows {
-            ops.push(*op);
-            for (datum, builder) in row.iter().zip_eq_debug(array_builders.iter_mut()) {
-                builder.append(datum);
-            }
+            let none = builder.append_row(*op, row);
+            debug_assert!(none.is_none());
         }
 
-        let new_columns = array_builders
-            .into_iter()
-            .map(|builder| builder.finish().into())
-            .collect::<Vec<_>>();
-        StreamChunk::new(ops, new_columns)
+        builder.take().expect("empty chunk")
```

Comment:
Prefer it to be type-safe, and return a None instead. But I think it will lead to large amount of refactoring, so we shouldn't refactor in this PR.

I guess subsequently we will refactor callers to use `StreamChunkBuilder` instead.

Reply:
I have reviewed all use cases for these functions and found all of them correctly handle the empty case before calling this function. I guess the unit tests run into some edge cases and will check then.